File: Abstract Factory/Exercicios/Ex2_Loja/Ex2.java
/**
 * PADRÃO: ABSTRACT FACTORY (FÁBRICA ABSTRATA)
 * 
 * Exercício: Sistema de Pagamentos de Loja.
 * O objetivo é garantir que a forma de pagamento (Cartão ou Boleto) 
 * seja processada pelo mesmo banco (Família Banco A ou Banco B).
 */

// 1. PRODUTOS ABSTRATOS
abstract class CartaoCredito {
    public abstract String pagar(double valor);
}

abstract class Boleto {
    public abstract String pagar(double valor);
}

// 2. PRODUTOS CONCRETOS - FAMÍLIA BANCO A
class CartaoBancoA extends CartaoCredito {
    @Override
    public String pagar(double valor) {
        return "Banco A: Cobrança de R$ " + String.format("%.2f", valor) + " no cartão realizada.";
    }
}

class BoletoBancoA extends Boleto {
    @Override
    public String pagar(double valor) {
        return "Banco A: Boleto de R$ " + String.format("%.2f", valor) + " gerado com sucesso.";
    }
}

// 2. PRODUTOS CONCRETOS - FAMÍLIA BANCO B
class CartaoBancoB extends CartaoCredito {
    @Override
    public String pagar(double valor) {
        return "Banco B: Pagamento de R$ " + String.format("%.2f", valor) + " aprovado no cartão!";
    }
}

class BoletoBancoB extends Boleto {
    @Override
    public String pagar(double valor) {
        return "Banco B: Aqui está seu boleto de R$ " + String.format("%.2f", valor);
    }
}

// 3. FABRICA ABSTRATA
abstract class FabricaBanco {
    public abstract CartaoCredito criarCartao();
    public abstract Boleto criarBoleto();
}

// 4. FABRICAS CONCRETAS
class FabricaBancoA extends FabricaBanco {
    @Override
    public CartaoCredito criarCartao() { return new CartaoBancoA(); }
    @Override
    public Boleto criarBoleto() { return new BoletoBancoA(); }
}

class FabricaBancoB extends FabricaBanco {
    @Override
    public CartaoCredito criarCartao() { return new CartaoBancoB(); }
    @Override
    public Boleto criarBoleto() { return new BoletoBancoB(); }
}

// 5. CLIENTE
public class Ex2 {
    public static void finalizarCompra(FabricaBanco fabrica, double valor) {
        // O cliente trabalha com as abstrações fornecidas pela fábrica
        CartaoCredito cartao = fabrica.criarCartao();
        Boleto boleto = fabrica.criarBoleto();

        System.out.println(" -> " + cartao.pagar(valor));
        System.out.println(" -> " + boleto.pagar(valor));
        System.out.println(" ----------------------------");
    }

    public static void main(String[] args) {
        System.out.println("\n--- FINALIZANDO NO BANCO A ---");
        finalizarCompra(new FabricaBancoA(), 420.75);

        System.out.println("\n--- FINALIZANDO NO BANCO B ---");
        finalizarCompra(new FabricaBancoB(), 189.30);
    }
}

File: Abstract Factory/Exercicios/Ex2_Loja/Ex2.py
"""
PADRÃO: ABSTRACT FACTORY (FÁBRICA ABSTRATA)

Exercício: Sistema de Pagamentos de Loja.
Neste exemplo, o Abstract Factory é usado para garantir a consistência 
entre os diferentes meios de pagamento oferecidos por cada banco.
"""

from abc import ABC, abstractmethod

# 1. PRODUTOS ABSTRATOS
class CartaoCredito(ABC):
    @abstractmethod
    def pagar(self, valor):
        pass

class Boleto(ABC):
    @abstractmethod
    def pagar(self, valor):
        pass

# 2. PRODUTOS CONCRETOS - FAMÍLIA BANCO A
class CartaoBancoA(CartaoCredito):
    def pagar(self, valor):
        return f"Banco A: Cobrança de R$ {valor:.2f} no cartão realizada."

class BoletoBancoA(Boleto):
    def pagar(self, valor):
        return f"Banco A: Boleto de R$ {valor:.2f} gerado com sucesso."

# 2. PRODUTOS CONCRETOS - FAMÍLIA BANCO B
class CartaoBancoB(CartaoCredito):
    def pagar(self, valor):
        return f"Banco B: Pagamento de R$ {valor:.2f} aprovado no cartão!"

class BoletoBancoB(Boleto):
    def pagar(self, valor):
        return f"Banco B: Aqui está seu boleto de R$ {valor:.2f}"

# 3. FABRICA ABSTRATA
class FabricaBanco(ABC):
    @abstractmethod
    def criar_cartao(self) -> CartaoCredito:
        pass
    
    @abstractmethod
    def criar_boleto(self) -> Boleto:
        pass

# 4. FABRICAS CONCRETAS
class FabricaBancoA(FabricaBanco):
    def criar_cartao(self):
        return CartaoBancoA()
    def criar_boleto(self):
        return BoletoBancoA()

class FabricaBancoB(FabricaBanco):
    def criar_cartao(self):
        return CartaoBancoB()
    def criar_boleto(self):
        return BoletoBancoB()

# 5. CLIENTE
def finalizar_compra(fabrica: FabricaBanco, valor: float):
    # O cliente não conhece as classes concretas, apenas a interface da fábrica
    cartao = fabrica.criar_cartao()
    boleto = fabrica.criar_boleto()

    print(f" -> {cartao.pagar(valor)}")
    print(f" -> {boleto.pagar(valor)}")
    print(" ----------------------------")

# Execução
if __name__ == "__main__":
    print("\n--- FINALIZANDO NO BANCO A ---")
    finalizar_compra(FabricaBancoA(), 420.75)

    print("\n--- FINALIZANDO NO BANCO B ---")
    finalizar_compra(FabricaBancoB(), 189.30)
